### PR TITLE
Add explicit verification for External/Internal IPv6 subnet

### PR DESCRIPTION
**Pull request by panslava**
- Before syncing IPv6 Services, verify if Service subnet has required IPv6 ranges (INTERNAL for ILB and EXTERNAL for NetLB)
- Add unit tests

### CODE REVIEW

File: pkg/loadbalancers/l4netlb_test.go [hunk 1]
```diff
@@ -583,37 +582,64 @@ func TestEnsureIPv6ExternalLoadBalancerCustomSubnet(t *testing.T) {
 	svc := test.NewL4NetLBRBSService(8080)
 	svc.Spec.IPFamilies = []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol}
 	l4NetLB, err := setupNetLBDualStackTestService(svc, nodeNames)
-	if err != nil {
-		t.Fatal(err)
-	}
```
[the remark this answers is not in the document]
Comment (cezarygerard) [reply pkg/loadbalancers/l4netlb_test.go:585]: test passes without them

File: pkg/loadbalancers/l4.go [hunk 2]
```diff
@@ -302,6 +319,21 @@ func (l4 *L4) EnsureInternalLoadBalancer(nodeNames []string, svc *corev1.Service
...
+			result.Error = err
+			return result
+		}
+		if !hasIPv6SubnetRange {
+			klog.Infof("Subnet %s for IPv6 Service %s/%s does not have internal IPv6 ranges", subnetName, l4.Service.Namespace, l4.Service.Name)
+			result.Error = utils.NewUserError(fmt.Errorf("subnet %s does not have internal IPv6 ranges, required for IPv6 Service", subnetName))
+			return result
+		}
+	}
+
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4.go:336]: this code is never unit tested

File: pkg/l4lb/l4controller_test.go [hunk 3]
```diff
@@ -568,6 +569,18 @@ func TestCreateDeleteDualStackService(t *testing.T) {
 				t.Errorf("Error getting L4 ILB latency metrics err: %v", err)
 			}
 			newSvc := test.NewL4ILBDualStackService(8080, api_v1.ProtocolTCP, tc.ipFamilies, api_v1.ServiceExternalTrafficPolicyTypeCluster)
+
+			// Create cluster subnet with Internal IPV6 range. Mock GCE uses subnet with empty string name.
```
Comment (cezarygerard) [inline pkg/l4lb/l4controller_test.go:573]: move this setup to helper func

File: pkg/l4lb/l4netlbcontroller_test.go [hunk 4]
```diff
@@ -1569,6 +1569,18 @@ func TestCreateDeleteDualStackNetLBService(t *testing.T) {
...
+			subnetKey := meta.RegionalKey(clusterSubnetName, controller.ctx.Cloud.Region())
+			subnetToCreate := &compute.Subnetwork{
+				Ipv6AccessType: "EXTERNAL",
+				StackType:      "IPV4_IPV6",
+			}
+			err := controller.ctx.Cloud.Compute().(*cloud.MockGCE).Subnetworks().Insert(context.TODO(), subnetKey, subnetToCreate)
+			if err != nil {
+				t.Fatal(err)
+			}
+
```
Comment (cezarygerard) [inline pkg/l4lb/l4netlbcontroller_test.go:1583]: move this setup to helper func

**Comment by cezarygerard** [comment]
good!

File: pkg/loadbalancers/l4netlb.go [hunk 5]
```diff
@@ -134,6 +134,21 @@ func (l4netlb *L4NetLB) EnsureFrontend(nodeNames []string, svc *corev1.Service)
...
 
+	// If service requires IPv6 LoadBalancer -- verify that Subnet with External IPv6 ranges is used.
+	if l4netlb.enableDualStack && utils.NeedsIPv6(svc) {
+		subnetName := l4netlb.ipv6SubnetName()
+		hasIPv6SubnetRange, err := utils.SubnetHasIPv6Range(l4netlb.cloud, subnetName, subnetExternalIPv6AccessType)
+		if err != nil {
+			result.Error = err
+			return result
+		}
+		if !hasIPv6SubnetRange {
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlb.go:145]: discussed offline, errors are emitted to users, so we don't need a custom event

added comment

File: pkg/l4lb/l4controller_test.go (hunk 3 above)
Comment (panslava) [reply pkg/l4lb/l4controller_test.go]: done

File: pkg/l4lb/l4netlbcontroller_test.go (hunk 4 above)
Comment (panslava) [reply pkg/l4lb/l4netlbcontroller_test.go]: done

File: pkg/loadbalancers/l4.go [hunk 6]
```diff
@@ -302,6 +319,21 @@ func (l4 *L4) EnsureInternalLoadBalancer(nodeNames []string, svc *corev1.Service
 		result.SyncType = SyncTypeUpdate
 	}
 
+	// If service requires IPv6 LoadBalancer -- verify that Subnet with Internal IPv6 ranges is used.
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4.go:322]: moved to private func

File: pkg/loadbalancers/l4.go (hunk 2 above)
Comment (panslava) [reply pkg/loadbalancers/l4.go]: added test TestILBUserErrorOnExternalIPv6Subnet

File: pkg/loadbalancers/l4netlb.go [hunk 7]
```diff
@@ -134,6 +134,21 @@ func (l4netlb *L4NetLB) EnsureFrontend(nodeNames []string, svc *corev1.Service)
...
+			klog.Infof("Subnet %s for IPv6 Service %s/%s does not have external IPv6 ranges", subnetName, l4netlb.Service.Namespace, l4netlb.Service.Name)
+			result.Error = utils.NewUserError(fmt.Errorf("subnet %s does not have external IPv6 ranges, required for IPv6 Service", subnetName))
+			return result
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlb.go]: done

File: pkg/loadbalancers/l4netlb_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlb_test.go]: actually I think this func should be "must", changed it

**Comment by panslava** [comment]
@cezarygerard ready for review

take a look please

File: pkg/loadbalancers/l4netlb_test.go [hunk 8]
```diff
@@ -583,37 +582,64 @@ func TestEnsureIPv6ExternalLoadBalancerCustomSubnet(t *testing.T) {
...
+		t.Fatalf("Failed to create subnet, error: %v", err)
+	}
 	svc.Annotations[annotations.CustomSubnetAnnotationKey] = "another-subnet"
 	result = l4NetLB.EnsureFrontend(nodeNames, svc)
 	if result.Error != nil {
-		t.Errorf("Failed to ensure loadBalancer, err %v", result.Error)
+		t.Fatalf("Failed to ensure loadBalancer, err %v", result.Error)
 	}
 	svc.Annotations = result.Annotations
 	svc.Annotations[annotations.CustomSubnetAnnotationKey] = "another-subnet"
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlb_test.go]: added comment

**Pull request merged**